arrow/obsidian.c

```diff
@@ -175,6 +175,18 @@ evolve_result evolve(Info *info, double duration, int64_t *state, double *rates)
       total += propensities[reaction];
     }
 
+    if (isnan(total)) {
+      printf("failed simulation: total propensity is NAN\n");
+      int max_reaction = 0;
+      for (reaction = 0; reaction < reactions_count; reaction++) {
+        if (propensities[reaction] > propensities[max_reaction]) {
+          max_reaction = reaction;
+        }
+      }
+      printf("largest reaction is %d at %f\n", max_reaction, propensities[max_reaction]);
+      total = 0.0;
```

Comment:
It could be more helpful for debugging if this returned the `NaN` [<-- note the capitalization] so the caller or the Python wrapper can check for that and raise an exception instead of printing to stdout without context. Also dropping the NaN loses the info on which kind of NaN it was.

BTW it might be able to avoid the overflow sometimes by using extended precision `long double` math and/or rearranging the order of operations to interleave multiplication and division.

Reply:
Hey @1fish2, thanks for this! It is a good question how to best handle this kind of failure. 

> It could be more helpful for debugging if this returned the NaN

The `NaN` arises in an intermediate value (`total`) which is not returned, but instead guides the rest of the computation. `StochasticSystem` returns a dictionary of five values:

* steps
* time
* events
* occurrences
* outcome

The current consideration is that the failure could have happened many steps into the simulation, which did produce reasonable results until that failure. Do we throw those results away? An argument could be made there, for now I am actually returning the useful parts of the results in case that is actually useful to the client. 

The main concern I think you raise that needs to be solved is that we want to communicate the failure back to the client in a way they can respond to programmatically, rather than the current "triage" message, which is more of a suggestion to fix the offending reaction and run it again. For this we could add another key to the result, like `status` perhaps? That describes the result, kind of like a bash exit code (0 for success, other numbers for failures). This would take more work to thread that result back through the c/cython machinery, but could be worth it if there was some meaningful action the client could take during the run of the simulation. Maybe just aborting would be most correct, but we could leave that up to the client. 

Currently I don't see demand for that feature (we are going to be using it to identify the offending reaction(s) and decompose the stoichiometry into subreactions) but for completeness/correctness sake it makes sense to provide.

Reply:
> BTW it might be able to avoid the overflow sometimes by using extended precision long double math and/or rearranging the order of operations to interleave multiplication and division.

We are currently using `float64` for the propensity calculations. I don't see a higher-bit data type here, besides `complex128` which just uses two `float64` for its two components: https://docs.scipy.org/doc/numpy-1.15.0/user/basics.types.html 

Are you referring to the 80-bit extended double precision? https://en.wikipedia.org/wiki/Extended_precision

I saw some references to 128-bit quadruple precision values, but they seem to be compiler dependent. 

> and/or rearranging the order of operations to interleave multiplication and division.

This overflow is happening in a loop while adding together thousands of large numbers. @tahorst suggested projecting to log space, which is an interesting approach to look into.

Reply:
Indeed maybe this is a low priority. In that case let's file an Issue for later. If this printf message was buried in the logs of 100 sims, I'd be lucky to notice it and would have no clue how to track down what went wrong where or which output values and downstream generations were GIGO. Cython can raise an exception and that'd fix it.

I'm under the impression that `long double` in C is an 80-bit extended precision float.

An overflow from adding lots of large numbers must mean LOTS of numbers that're already near overflow. Doing it in log space is not easy and would reduce precision. It'd be easier to scale them down by say 1000 into a more reasonable range; and if the total will be divided by some factor anyway it's even easier to divide them by that before adding (rearranging the order of operations). Still maybe to fix in the future.

Reply:
> Cython can raise an exception and that'd fix it.

Okay, I made a stab at this: https://github.com/CovertLab/arrow/commit/da7bcb3ad7017953f46fff3b11fc198c6657ebeb, though I am currently getting a weird error when trying to add a new `status` field to the `evolve_result` struct, cython does not seem to find it, even though it is definitely there:

```
(arrow) spanglry@spanglry:~/Code/arrow$ make clean compile
USE_CYTHON=1 python setup.py build_ext --inplace
running build_ext
cythoning arrow/arrowhead.pyx to arrow/arrowhead.c

Error compiling Cython file:
------------------------------------------------------------
...
        Return None or a tuple (steps, time, events, outcome).
        """
        # TODO(jerry): Check the state[] array size?

        evolved = obsidian.evolve(&self.info, duration, &state[0], &rates[0])
        cdef int status = evolved.status
                                ^
------------------------------------------------------------

arrow/arrowhead.pyx:108:33: Object of type 'evolve_result' has no attribute 'status'
building 'arrow.arrowhead' extension
C compiler: gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC

compile options: '-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -Iarrow -I/home/spanglry/Code/arrow/arrow -I/home/spanglry/.pyenv/versions/arrow/lib/python2.7/site-packages/numpy/core/include -I/home/spanglry/.pyenv/versions/2.7.15/include/python2.7 -c'
gcc: arrow/arrowhead.c
arrow/arrowhead.c:1:2: error: #error Do not use this file, it is the result of a failed Cython compilation.
 #error Do not use this file, it is the result of a failed Cython compilation.
  ^~~~~
arrow/arrowhead.c:1:2: error: #error Do not use this file, it is the result of a failed Cython compilation.
 #error Do not use this file, it is the result of a failed Cython compilation.
  ^~~~~
error: Command "gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -Iarrow -I/home/spanglry/Code/arrow/arrow -I/home/spanglry/.pyenv/versions/arrow/lib/python2.7/site-packages/numpy/core/include -I/home/spanglry/.pyenv/versions/2.7.15/include/python2.7 -c arrow/arrowhead.c -o build/temp.linux-x86_64-2.7/arrow/arrowhead.o" failed with exit status 1
Makefile:10: recipe for target 'compile' failed
make: *** [compile] Error 1

```

Any thoughts on that one?

Reply:
On the `status` branch if you want to check it out. I've tried compiling just the obsidian stuff first, then compiling the cython (trying to "reset" the header state??) but I get the same result. I have no idea why it won't find the new header? Very strange. 

Perhaps it is finding the old header somewhere else on the system?